src/esw/can_node/can_node.cpp

```diff
@@ -0,0 +1,120 @@
+#include "can_node.hpp"
+
+#include <mutex>
+#include <nodelet/loader.h>
+#include <ros/init.h>
+#include <ros/names.h>
+#include <ros/this_node.h>
+#include <stdexcept>
+
+namespace mrover {
+
+    constexpr size_t CAN_ERROR_BIT_INDEX = 29;
+    constexpr size_t CAN_RTR_BIT_INDEX = 30;
+    constexpr size_t CAN_EXTENDED_BIT_INDEX = 31;
+
+    void CanNodelet::onInit() {
+        try {
+            mNh = getMTNodeHandle();
+            mPnh = getMTPrivateNodeHandle();
+            mCanSubscriber = mNh.subscribe<CAN>("can_requests", 1, &CanNodelet::handleMessage, this);
+
+            mIsExtendedFrame = mNh.param<bool>("is_extended_frame", false);
+
+            if ((mSocket = socket(PF_CAN, SOCK_RAW, CAN_RAW)) < 0) {
+                throw std::runtime_error("Failed to open socket");
+            }
+
+            ifreq ifr{};
+            const char* interfaceName = "can0";
+            strcpy(ifr.ifr_name, interfaceName);
+            ioctl(mSocket, SIOCGIFINDEX, &ifr);
+
+            sockaddr_can addr{
+                    .can_family = AF_CAN,
+                    .can_ifindex = ifr.ifr_ifindex,
+            };
+
+
+            if (bind(mSocket, reinterpret_cast<sockaddr*>(&addr), sizeof(addr)) < 0) {
+                throw std::runtime_error("Failed to bind to socket");
+            }
+        } catch (std::exception const& exception) {
+            ROS_ERROR_STREAM(exception.what());
+            ros::requestShutdown();
+        }
+    }
+
+    void CanNodelet::setBus(uint8_t bus) {
+        this->mBus = bus;
+    }
+
+    // todo(owen): while loop until nbytes == size of can frame
+    void CanNodelet::sendFrame() const {
+        size_t nbytes = write(mSocket, &mFrame, sizeof(can_frame));
```

Comment:
Put `write` in a loop until it writes `sizeof(can_frame)`.

It is possible for write to process chunks in multiple requests.